### PR TITLE
Improve TDex flow (discovery & tdex registry)

### LINKED ISSUE #4: Use Discoverer class instead of hardcoding two providers

### PR DESCRIPTION
**Pull request by louisinger**
This PR improves the tdex flow in the app:

- fetch providers (and markets) from registry (use a new Svelte store: `tdexstore.ts`).
- handle all assets, fetch Infos from explorer (new svelte store: `coinstore.ts`).
- use Discoverer to select best orders from providers (`Trade.svelte`).
- use the marina UTXO events to get and delete utxos instead of `marina.getCoins` call.

it reuses the functions of tdex-app (see `utils/tdex.ts` file)

Note that this PR does not add support for testnet or regtest (does it make sense on dex.vulpem.com?). And do not add any new "visual elements" like the number of providers/markets available etc... (may be useful ??).

it closes #4 

@tiero @bordalix please review

### CODE REVIEW

File: src/components/FiatValue.svelte [hunk 1]
```diff
@@ -25,8 +30,8 @@
   let promise;
   let visible;
 
-  $: id = CoinGeckoId[coin];
-  $: currency = CoinGeckoId[fiat];
+  $: id = CoinGeckoId[coin.assetHash];
+  $: currency = CoinGeckoId[fiat.toString()];
   $: promise = getValue(id, amount, currency);
```
Comment (bordalix) [inline src/components/FiatValue.svelte:35]: Nice catch on the `toString()`

File: src/constants/assets.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  },
+  {
+    ticker: 'LCAD',
+    assetHash: LIQUID_CAD,
+    precision: 8,
+    name: 'Liquid CAD',
+  },
+  {
+    ticker: 'BTSE',
+    assetHash: 'b00b0ff0b11ebd47f7c6f57614c046dbbd204e84bf01178baf2be3713a206eb7',
```
Comment (bordalix) [inline src/constants/assets.ts:25]: Why don't have this on constants/index.js?

Comment (louisinger) [reply src/constants/assets.ts:25]: Mostly due to copy/paste I think, u right I'll move it :+1:

File: src/containers/Trade.svelte [hunk 3]
```diff
@@ -180,38 +186,24 @@
...
+          sats: amountToBeSentInSatoshis.toNumber(),
+          asset: pair.send.assetHash,
+        },
+        bestOrder,
+        $marinaStore.utxos,
+        EXPLORER
+      );
 
       tradeStatus = TradeStatus.COMPLETED;
     } catch (e) {
```
Comment (bordalix) [inline src/containers/Trade.svelte:209]: There is an issue: are losing specificity on the error shown to user.

If the user has no available utxo, previously the error shown was "You need at least one utxo to trade", which is good, but now the error shown to the user is "error while making trade", which is worst.

The same thing occurs when there are no orders available.

File: src/stores/tdexstore.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  markets: {},
+};
+
+export const tdexStore = readable<TdexStore>(initialTdexState, function(set: Subscriber<TdexStore>) {
+  const fetchProvidersAndMarketsAndSet = async () => {
+    const providers = await getProvidersFromRegistry();
+    const markets = await getMarketsForProviders(providers);
+    set({ providers, markets });
+  }
+  console.log('Fetching providers and markets...');
```
Comment (bordalix) [inline src/stores/tdexstore.ts:24]: Should we remove this console.log()?

File: src/utils/tdex.ts [hunk 5]
```diff
@@ -1,15 +1,167 @@
...
+      asset: known.asset,
+      market: order.market,
+      identity
+    }
+    const promise = order.type === TradeType.BUY ? trade.buy(args) : trade.sell(args);
+    const txid = await promise;
+    return txid;
+  } catch (e) {
+    console.error(e);
+    throw new Error('error while making trade');
```
Comment (bordalix) [inline src/utils/tdex.ts:165]: Why not throw the `e.message` error itself?

**Review by bordalix — COMMENTED** [review]
ACK.

My only issue is not showing the user the true reason for the error.

**Review by bordalix — COMMENTED** [review]
Yes, I also prefer to have all the constants on the same file.

File: src/utils/tdex.ts (hunk 5 above)
Comment (louisinger) [reply src/utils/tdex.ts]: In fact, tdex-sdk does not return very explicit error messages (more about this [here](https://github.com/tdex-network/tdex-app/issues/443) and [here](https://github.com/tdex-network/tdex-sdk/issues/121)) that's why we wrap the buy/sell call with a try-catch statement in order to return a more "human-readable error msg" in tdex-app. 
However, looking at master, it looks like we just return the raw SDK errors so I've just dropped the try-catch statement and let `Trade.svelte` handle errors.

File: src/stores/coinstore.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+  subscribe((v) => state = v);
+
+  return {
+    subscribe,
+    getCoin: (assetHash: string) => state.coins.find(c => c.assetHash === assetHash) || assetHashToCoin(assetHash),
+    updateWithAssets: async (assets: string[]) => {
+      for (const asset of assets) {
+        if (!state.coins.find(c => c.assetHash === asset)) {
+          const newCoin = await getAssetData(asset, EXPLORER);
```
Comment (tiero) [inline src/stores/coinstore.ts:27]: what's happen if we got an error here

File: src/stores/store.ts [hunk 7]
```diff
@@ -1,15 +1,17 @@
...
 
-
 export interface MarinaStore {
   installed: boolean;
   enabled: boolean;
+  utxos: Utxo[];
 }
 
-const initialState = { installed: false, enabled: false, network:"liquid" }
+const initialState = { installed: false, enabled: false, network: "liquid", utxos: [] };
```
Comment (tiero) [inline src/stores/store.ts:10]: I would keep utxos in a different store (potentially we could use `svelte-marina-button` which handles this marinaStore as a cache for connection status only

File: src/stores/coinstore.ts (hunk 6 above)
Comment (louisinger) [reply src/stores/coinstore.ts:27]: `getCoin` won't find the "Coin" associated to your asset hash. So `assetHashToCoin(assetHash)` (which is basically default values) will be used

Comment (louisinger) [reply src/stores/coinstore.ts:27]: Note that if the call in getAssetData fails, we also catch the error and return `assetHashToCoin(assetHash)`

File: src/containers/Connect.svelte [hunk 8]
```diff
@@ -43,16 +44,48 @@
...
+      marina.on('DISABLED', () => {
+        notifyDisabled();
+      })
+    );
+
+    listeners.push(
+      marina.on('NEW_UTXO', (u: Utxo) => {
+        marinaStore.update((s: MarinaStore) => ({
+          ...s,
+          utxos: [...s.utxos, u],
```
Comment (tiero) [inline src/containers/Connect.svelte:63]: not clear how we can get the coins already in marina (ie before we start listening) withotu calling the getCoins?

Comment (louisinger) [reply src/containers/Connect.svelte:63]: the `NEW_UTXO` event is **relative to the injected page** it means that "NEW" doesn't mean "new in marina" but "never sent to web page". So if we subscribe to `NEW_UTXO` event type, we'll receive first all the "existing" unspents directly. Maybe this is not clear in doc I think

File: src/stores/store.ts (hunk 7 above)
Comment (louisinger) [reply src/stores/store.ts]: done

File: src/containers/Connect.svelte (hunk 8 above)
Comment (tiero) [reply src/containers/Connect.svelte]: ah totally missed this feature. Yes will need to make it more clear

**Review by tiero — APPROVED** [review]

**Pull request merged**